plutus-core/plutus-core/src/PlutusCore/Builtin/Convert.hs

```diff
@@ -0,0 +1,526 @@
+-- editorconfig-checker-disable-file
+
+{-# LANGUAGE OverloadedStrings #-}
+
+-- | Implementations for conversion primops from 'Integer' to 'ByteString' and back again.
+module PlutusCore.Builtin.Convert (
+  -- Wrappers
+  integerToByteStringWrapper,
+  byteStringToIntegerWrapper,
+  -- Implementation details
+  IntegerToByteStringError(..),
+  integerToByteString,
+  byteStringToInteger
+  ) where
+
+import ByteString.StrictBuilder (Builder)
+import ByteString.StrictBuilder qualified as Builder
+import Control.Monad (guard)
+import Data.Bits (unsafeShiftL, unsafeShiftR, (.|.))
+import Data.ByteString (ByteString)
+import Data.ByteString qualified as BS
+import Data.Text (pack)
+import Data.Word (Word64)
+import GHC.ByteOrder (ByteOrder (BigEndian, LittleEndian))
+import PlutusCore.Builtin.Emitter (Emitter, emit)
+import PlutusCore.Evaluation.Result (EvaluationResult (EvaluationFailure))
+
+-- | Wrapper for 'integerToByteString' to make it more convenient to define as a builtin.
+integerToByteStringWrapper ::
+  Bool -> Integer -> Integer -> Emitter (EvaluationResult ByteString)
+integerToByteStringWrapper endiannessArg paddingArg input =
+  let endianness = if endiannessArg then BigEndian else LittleEndian in
+    -- We use fromIntegral here, despite advice to the contrary in general when defining builtin
+    -- denotations. For why we do this (and why it's both inevitable and not really a concern
+    -- anyway), see Note [fromIntegral and padding arguments].
+    case integerToByteString (fromIntegral (max 0 paddingArg)) endianness input of
+      Left err -> case err of
+        NegativeInput -> do
+          emit "builtinIntegerToByteString: cannot convert negative Integer"
+          emit $ "Input: " <> (pack . show $ input)
```

Comment:
The only way I see to fix this would be to not display the integer that caused the failure (unless I'm misunderstanding), which would make debugging much more annoying. Additionally, if we have extremely large (in terms of absolute value) inputs, wouldn't costing catch this anyway? @kwxm - could you confirm I'm not missing something?

Happy to add a comment.

Reply:
Yeah, I think it's the right thing to do, it's just _slightly_ subtle. In fact I think it slightly relies on `builtinIntegerToByteString` being costed proportionally to the size of the integer.

Perhaps:
```suggestion
          -- This does work proportional to the size of the integer, but we're in a failing case,
          -- an the user has paid for work proportional to the size of the integer already.
          emit $ "Input: " <> (pack . show $ input)
```